samples/java-eventsourced-shopping-cart/src/main/java/com/example/shoppingcart/domain/ShoppingCartImpl.java

```diff
@@ -16,94 +16,111 @@
+                            .setName(command.getName())
+                            .setQuantity(command.getQuantity())
+                            .build())
+            .build();
+
+    return effectBuilder
+            .emitEvent(event)
+            .thenReply(newState -> Empty.getDefaultInstance());
   }
 
   @Override
-  protected Empty removeItem(ShoppingCartApi.RemoveLineItem item, CommandContext ctx) {
-    if (!cart.containsKey(item.getProductId())) {
-      throw ctx.fail(
-          "Cannot remove item " + item.getProductId() + " because it is not in the cart.");
+  public Effect<Empty> removeItem(
+          ShoppingCartApi.RemoveLineItem command,
+          ShoppingCartDomain.Cart currentState,
+          EventSourcedEntityEffect.Builder<Empty, ShoppingCartDomain.Cart> effectBuilder,
+          CommandContext ctx) {
+    if (findItemByProductId(currentState, command.getProductId()).isEmpty()) {
+      return effectBuilder.failure(
+          "Cannot remove item " + command.getProductId() + " because it is not in the cart.");
     }
-    ctx.emit(ShoppingCartDomain.ItemRemoved.newBuilder().setProductId(item.getProductId()).build());
-    return Empty.getDefaultInstance();
+
+    ShoppingCartDomain.ItemRemoved event =
+            ShoppingCartDomain.ItemRemoved.newBuilder().setProductId(command.getProductId()).build();
+
+    return effectBuilder
+            .emitEvent(event)
+            .thenReply(newState -> Empty.getDefaultInstance());
   }
 
   @Override
-  protected ShoppingCartApi.Cart getCart(ShoppingCartApi.GetShoppingCart command, CommandContext ctx) {
-    return ShoppingCartApi.Cart.newBuilder().addAllItems(cart.values()).build();
+  public Effect<ShoppingCartApi.Cart> getCart(
+          ShoppingCartApi.GetShoppingCart command,
+          ShoppingCartDomain.Cart currentState,
+          EventSourcedEntityEffect.Builder<ShoppingCartApi.Cart, ShoppingCartDomain.Cart> effectBuilder,
+          CommandContext ctx) {
+    List<ShoppingCartApi.LineItem> apiItems =
+            currentState.getItemsList().stream()
+            .map(this::convert)
+            .sorted(Comparator.comparing(ShoppingCartApi.LineItem::getProductId))
+            .collect(Collectors.toList());
+    ShoppingCartApi.Cart apiCart = ShoppingCartApi.Cart.newBuilder().addAllItems(apiItems).build();
+    return effectBuilder.message(apiCart);
   }
 
   @Override
-  public void itemAdded(ShoppingCartDomain.ItemAdded itemAdded) {
-    ShoppingCartApi.LineItem item = cart.get(itemAdded.getItem().getProductId());
-    if (item == null) {
-      item = convert(itemAdded.getItem());
-    } else {
-      item =
-          item.toBuilder()
-              .setQuantity(item.getQuantity() + itemAdded.getItem().getQuantity())
-              .build();
-    }
-    cart.put(item.getProductId(), item);
+  protected ShoppingCartDomain.Cart itemAdded(ShoppingCartDomain.ItemAdded itemAdded, ShoppingCartDomain.Cart currentState) {
```

Comment:
Yes, I think it reads better with state is first. In my brain I read it as: given this `State`, what should happen if we apply this `Command`. 

Which also reads as, given this `State` what is the command handler (`Cmd => Effect`) we have for it.

Same reasoning applies event handlers. 

So, yeah. State first all the way down.

Reply:
State first would probably read better. JS SDK has command, state, context. Probably so that handlers can just have the command if current state is not needed.

Reply:
I have changed the order, state first: 9017436